realm/realm-library/src/main/java/io/realm/internal/android/JsonUtils.java

```diff
@@ -37,13 +41,26 @@
      */
     public static Date stringToDate(String date) {
         if (date == null || date.length() == 0) return null;
+
+        // Check for JSON date
         Matcher matcher = jsonDate.matcher(date);
         if (matcher.find()) {
             String dateMatch = matcher.group(1);
             return new Date(Long.parseLong(dateMatch));
-        } else {
+        }
+
+        // Check for millisecond based date
+        if(numericOnly.matcher(date).matches()) {
             return new Date(Long.parseLong(date));
         }
+
+        // Try for ISO8601 date
+        try {
+            parsePosition.setIndex(0); // reset the position each time
+            return ISO8601Utils.parse(date, parsePosition);
+        } catch (ParseException ignored) {
+            return null;
```

Comment:
Ah didn't notice that `ParseException` was a checked exception. Yes, catching and throwing a runtime exception is a lot better.

I am not entirely sure what would be best. `RealmException` or just `RuntimeException`?  I am probably leaning towards just `RuntimeException` as we have been removing use of `RealmException` elsewhere, but either would be fine I think.

If we used `RealmException` we should probably convert the NumberFormatException to that as well though.